cmd/snap-device-helper/snap-device-helper.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include <errno.h>
+#include <libgen.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "../libsnap-confine-private/cleanup-funcs.h"
+#include "../libsnap-confine-private/device-cgroup-support.h"
+#include "../libsnap-confine-private/snap.h"
+#include "../libsnap-confine-private/string-utils.h"
+#include "../libsnap-confine-private/utils.h"
+
+#include "snap-device-helper.h"
+
+static unsigned long must_strtoul(char *str) {
+    char *end = str;
+    unsigned long val = strtoul(str, &end, 10);
+    if (*end != '\0') {
+        die("malformed number \"%s\"", str);
+    }
+    return val;
+}
+
+/* udev_to_security_tag converts a udev tag (snap_foo_bar) to security tag
+ * (snap.foo.bar) */
```

Comment:
this should also probably mention the assumptions like that the string starts with `snap_` etc